src/target/riscv/riscv-013.c

```diff
@@ -1744,6 +1744,18 @@ static int examine(struct target *target)
 		r->mtopi_readable = false;
 	}
 
+	riscv_reg_t dcsr;
+	if (register_read_direct(target, &dcsr, GDB_REGNO_DCSR) != ERROR_OK)
+		return ERROR_FAIL;
+	dcsr = set_field(dcsr, CSR_DCSR_STEP, 0);
+	dcsr = set_field(dcsr, CSR_DCSR_EBREAKM, riscv_ebreakm);
+	dcsr = set_field(dcsr, CSR_DCSR_EBREAKS, riscv_ebreaks);
+	dcsr = set_field(dcsr, CSR_DCSR_EBREAKU, riscv_ebreaku);
+	dcsr = set_field(dcsr, CSR_DCSR_EBREAKVS, riscv_ebreaku);
+	dcsr = set_field(dcsr, CSR_DCSR_EBREAKVU, riscv_ebreaku);
+	if (register_write_direct(target, GDB_REGNO_DCSR, dcsr) != ERROR_OK)
+		return ERROR_FAIL;
+
```

Comment:
The same operation is performed in `riscv013_on_step_or_resume()`, so the code can be factored out to a single helper function.

The setting of `dcsr.ebreak*` bits should be performed also in `deassert_reset`. I have recorded it as an issue: https://github.com/riscv/riscv-openocd/issues/853 .

Reply:
I've refactored. Let's leave the deassert fix for another PR.

Reply:
Sure, the setting of `dcsr.ebreak* ` is a topic for a separate PR. I've mentioned it just for completeness :) Thanks.